client/scripts/mifarePlus.lua

```diff
@@ -0,0 +1,310 @@
+local cmds = require('commands')
+local lib14a = require('read14a')
+
+SIXTEEN_BYTES_ZEROS = "00000000000000000000000000000000"
+
+GETVERS_INIT = "0360" -- Begins the GetVersion command
+GETVERS_CONT = "03AF" -- Continues the GetVersion command
+POWEROFF = "OFF"
+WRITEPERSO = "03A8"
+COMMITPERSO = "03AA"
+AUTH_FIRST = "0370"
+AUTH_CONT = "0372"
+AUTH_NONFIRST = "0376"
+PREPAREPC = "03F0"
+PROXIMITYCHECK = "03F2"
+VERIFYPC = "03FD"
+READPLAINNOMACUNMACED = "0336"
+
+--- 
+-- This is only meant to be used when errors occur
+function oops(err)
+	print("ERROR: ",err)
+end
+
+---
+-- Used to send raw data to the firmware to subsequently forward the data to the card.
+function sendRaw(rawdata, crc, power)
+	print(("<sent>: 	  %s"):format(rawdata))
+
+	local flags = lib14a.ISO14A_COMMAND.ISO14A_RAW
+	if crc then
+		flags = flags + lib14a.ISO14A_COMMAND.ISO14A_APPEND_CRC
+	end
+	if power then
+		flags = flags + lib14a.ISO14A_COMMAND.ISO14A_NO_DISCONNECT
+	end
+
+	local command = Command:new{cmd = cmds.CMD_READER_ISO_14443a, 
+								arg1 = flags, -- Send raw
+								arg2 = string.len(rawdata) / 2, -- arg2 contains the length, which is half the length of the ASCII-string rawdata
+								data = rawdata}
+	local ignore_response = false
+	local result, err = lib14a.sendToDevice(command, ignore_response)
+	if result then
+		--unpack the first 4 parts of the result as longs, and the last as an extremely long string to later be cut down based on arg1, the number of bytes returned
+		local count,cmd,arg1,arg2,arg3,data = bin.unpack('LLLLH512',result)
+		returned_bytes = string.sub(data, 1, arg1 * 2)
+		print(("<recvd>: %s"):format(returned_bytes)) -- need to multiply by 2 because the hex digits are actually two bytes when they are strings
+		return returned_bytes
+	else
+		err = "Error sending the card raw data."
+		oops(err)
+	end
+end
+
+function writePerso()
+	-- Used to write any data, including the keys (Key A and Key B), for all the sectors.
+	-- writePerso() command parameters:
+	--		1 byte  - 0xA8 - Command Code
+	--		2 bytes - Address of the first block or key to be written to (40 blocks are numbered from 0x0000 to 0x00FF)
+	--		X bytes - The data bytes to be written, starting from the first block. Amount of data sent can be from 16 to 240 bytes in 16 byte increments. This allows
+	--				up to 15 blocks to be written at once.
+	-- response from PICC:
+	--		0x90 - OK
+	--		0x09 - targeted block is invalid for writes, i.e. block 0, which contains manufacturer data
+	--		0x0B - command invalid
+	--		0x0C - unexpected command length
+	
+
+	-- First, set all the data in the card to zeros. The keys, stored in the sector trailer block, are also set to zeros.
```

Comment:
The access bits will also be set to zeros? Definitely not a good idea!

Reply:
They can always be changed later, or someone can change the script. The important thing is that we explicitly set all the keys.

Reply:
What about this:

"**Remark**: With each memory access the internal logic verifies the format of the access conditions. If it detects a format violation the whole sector is irreversible blocked."

Reply:
Now that I read more about the access conditions, I see that you are right in saying that setting all the card data to 0 complicates things. Given that "On chip delivery the access conditions for the sector trailer and KeyA are predefined as transport configuration", I think it makes sense to just get rid of this code, so I will do so and leave the keys as the defaults.